src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -502,19 +500,49 @@ private static BoundExpression CreateAnonymousFunctionConversion(SyntaxNode synt
             // UNDONE: is converted to a delegate that does not match. What to surface then?
 
             var unboundLambda = (UnboundLambda)source;
-            var boundLambda = unboundLambda.Bind((NamedTypeSymbol)destination);
-            diagnostics.AddRange(boundLambda.Diagnostics);
+            var expr = createAnonymousFunctionConversion(syntax, unboundLambda, conversion, isCast, conversionGroup, destination, diagnostics);
 
-            return new BoundConversion(
-                syntax,
-                boundLambda,
-                conversion,
-                @checked: false,
-                explicitCastInCode: isCast,
-                conversionGroup,
-                constantValueOpt: ConstantValue.NotAvailable,
-                type: destination)
-            { WasCompilerGenerated = source.WasCompilerGenerated };
+            if (destination.SpecialType == SpecialType.System_Delegate || destination.IsNonGenericExpressionType())
+            {
+                CheckFeatureAvailability(syntax, MessageID.IDS_FeatureInferredDelegateType, diagnostics);
+                CompoundUseSiteInfo<AssemblySymbol> useSiteInfo = GetNewCompoundUseSiteInfo(diagnostics);
+                conversion = Conversions.ClassifyConversionFromExpression(expr, destination, ref useSiteInfo);
+                diagnostics.Add(syntax, useSiteInfo); // PROTOTYPE: Report use-site diagnostics.
+                return CreateConversion(syntax, expr, conversion, isCast, conversionGroup, destination, diagnostics);
+            }
+
+            return expr;
+
+            BoundConversion createAnonymousFunctionConversion(SyntaxNode syntax, UnboundLambda unboundLambda, Conversion conversion, bool isCast, ConversionGroup? conversionGroup, TypeSymbol destination, BindingDiagnosticBag diagnostics)
+            {
+#if DEBUG
+                if (destination.SpecialType != SpecialType.System_Delegate)
+                {
+                    _ = unboundLambda.Bind(Compilation.GetSpecialType(SpecialType.System_Delegate));
+                }
+#endif
+                var boundLambda = unboundLambda.Bind((NamedTypeSymbol)destination);
```

Comment:
Is there a reason you don't have the `destination` parameter just typed as `NamedTypeSymbol`?

Reply:
We could, although `destination` is assigned `boundLambda.Type` below so that would need to change too. Leaving as `TypeSymbol` seems slightly simpler.